moving_modules.md

```diff
@@ -0,0 +1,224 @@
+
+## Server-based Catalog Compilation
+
+As an admin, I should be able to compile catalogs for resources whose types are included in the locally installed puppet-agent package, so that I don't have to install modules I already have.
+
+## Server-based Catalog Application
+
+As an admin, if catalog compilation succeeds, then I want assurance that the catalog will be applied consistently across all agent versions regardless of which preinstalled modules are on each agent. For example, if the master has a newer version of a type, then all agents in that environment should use the same version of the type (and its provider) at catalog application time.
+
+## Module Updates
+
+As an admin, if there is a problem with a type/provider packaged with puppet-agent, I want to be able to install a newer version from the forge (for both serverless and server-based), so that I don't need to wait for Puppet to release a new puppet-agent build.
+
+ ## Puppet-Agent Updates
+
+As an admin, when the puppet-agent package is updated, I want to be able to use the new versions of types and providers that come with the new package, even if that means overwriting and deleting older preinstalled versions. However, installing a new puppet-agent package should not clobber modules I've installed via puppet module tool, r10k, etc.
+
+## Module Pinning
+
+As an admin, if I update the puppet-agent package, and it updates a preinstalled module, but the module introduces a regression, I want to easily install the older version of the module from the forge.  I don't want to be forced to downgrade puppet-agent packages, because that process introduces more risk.
+
+# File Path Changes
+
+Puppet's default `basemodulepath` includes two module directories visible to all environments:
+
+    /etc/puppetlabs/code/modules
+    /opt/puppetlabs/puppet/modules
+
+The first is where modules are typically installed to via puppet module tool, r10k, codemanager/filesync. The second path typically contains modules that are installed in PE environments, though there is nothing stopping users from manually installing modules there, e.g. `puppet module tool install puppetlabs-apt --target-dir /opt/puppetlabs/puppet/modules`. It is important that preinstalled modules do not use those same locations, otherwise, it will confuse package managers.
+
+We propose a new directory `/opt/puppetlabs/puppet/vendored_modules` to be created at puppet-agent installation time and containing all modules added to puppet-agent at packaging. The directory should be appended to the default `basemodulepath` so that the modules are available during compilation and application.
```

Comment:
how about `vendor_modules` for parallelism with ruby's `vendor_ruby` area?

Reply:
Good point, thoughts @branan?

Reply:
![neutral response](https://i.imgur.com/8LbUk7V.gif)